apis/apps/v1alpha1/cluster_types.go

```diff
@@ -293,12 +294,41 @@ type ReplicationMemberStatus struct {
 	Pod string `json:"pod"`
 }
 
+type SwitchoverCandidate struct {
```

Comment:
> is it a one-shot request or a proposed state to cluster?

SwitchoverCandidate is more like proposed state to cluster, It will switchover the components that meet the requirements through the reconcile process.

Reply:
the system needs to keep the requirements until when or what conditions matched?

Reply:
> the system needs to keep the requirements until when or what conditions matched?

Under normal circumstances, when the user modifies the switchoverCandidate, when the condition in the switchoverCandidate is inconsistent with the actual state in the system, we believe that the user actively triggered a switchover, and the system will execute the switchover action at this time. After the switchover is completed, the actual state of the system It should be consistent with switchoverCandidate, which is also the scenario in most cases. When reconcile detects that the state is consistent, it will not do anything until the next time the user starts switchover again.

Reply:
> the system needs to keep the requirements until when or what conditions matched?

However, in some extreme cases, the system may failover and cause the switchoverCandidate state to be inconsistent with the actual state of the system. In this scenario, the switchover is not triggered by the user. In this inconsistent state, reconcile will decide whether to synchronize the actual state in the system back to switchoverCandidate according to the user-defined switchoverCandidate.roleSync policy. If the user sets roleSync are not synchronized, this inconsistent state will continue until the user actively triggers switchover again.

Reply:
> 

it seems cleanup or delete the SwitchoverCadidate after a switchover is more reasonable than sync?

Reply:
@nashtsai what's "the design principles of k8s" here?

Reply:
@Y-Rookie the problems with sync are:
1. The SwitchoverCadidate will have different semantics after a synchronization, and the semantics are unclear, what's the meaning to maintain them all the time?
2. Maintain the synchronization with failover, it requires to keep consistency with the role info in Status,  which are duplicated and will introduce additional complexity